### PR TITLE
Fix for complex projections that include navigations

### LINKED ISSUE #3633: Query :: QuerySourceTracingExpressionVisitor doesn't prune EF.Property calls when looking for query sources
repro:

```
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.Linq;
using Microsoft.Data.Entity;

namespace Repro3633
{
    public class A
    {
        public int Id { get; set; }

        public int CId { get; set; }
        public virtual C C { get; set; }
    }

    public class C
    {
        public int Id { get; set; }

        public virtual ICollection<A> A { get; set; }
    }

    public class MyContext : DbContext
    {
        public DbSet<A> As { get; set; }
        public DbSet<C> Cs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Repro3633", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<A>().HasOne(p => p.C).WithMany(p => p.A).HasForeignKey(p => p.CId);
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                C c = new C();
                ctx.Cs.Add(c);

                A a = new A();
                a.C = c;
                a.CId = c.Id;
                ctx.As.Add(a);

                ctx.SaveChanges();
            }

            using (var ctx = new MyContext())
            {
                var query = from a in ctx.As.Include(a => a.C)
                            select new 
                            {
                                C = new 
                                {
                                    Id = a.C.Id
                                }
                            };

                var result = query.ToList();
            }
        }
    }
}

```

throws:

Unhandled Exception: System.ArgumentException: The given expression 'new DtoA() {C = new DtoC() {Id = Property([a], "CId")}}' does not contain the searched expression '[a]' in a nested NewExpression with member assignments or a MemberBindingExpression.
Parameter name: fullExpression
   at Remotion.Linq.Clauses.ExpressionVisitors.AccessorFindingExpressionVisitor.FindAccessorLambda(Expression searchedExpression, Expression fullExpression, ParameterExpression inputParameter)
   at Microsoft.Data.Entity.Query.EntityQueryModelVisitor.IncludeNavigations(QueryModel queryModel, IReadOnlyCollection`1 includeSpecifications) in D:\k\EntityFramework\src\EntityFramework.Core\Query\EntityQueryModelVisitor.cs:line 352
   at Microsoft.Data.Entity.Query.RelationalQueryModelVisitor.IncludeNavigations(QueryModel queryModel, IReadOnlyCollection`1 includeSpecifications) in D:\k\EntityFramework\src\EntityFramework.Relational\Query\RelationalQueryModelVisitor.cs:line 195
   at Microsoft.Data.Entity.Query.EntityQueryModelVisitor.IncludeNavigations(QueryModel queryModel) in D:\k\EntityFramework\src\EntityFramework.Core\Query\EntityQueryModelVisitor.cs:line 331
   at Microsoft.Data.Entity.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel) in D:\k\EntityFramework\src\EntityFramework.Core\Query\EntityQueryModelVisitor.cs:line 165
   at Microsoft.Data.Entity.Storage.Database.CompileQuery[TResult](QueryModel queryModel) in D:\k\EntityFramework\src\EntityFramework.Core\Storage\Database.cs:line 68
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Data.Entity.Query.Internal.QueryCompiler.<>c__DisplayClass18_0`1.<CompileQuery>b__0() in D:\k\EntityFramework\src\EntityFramework.Core\Query\Internal\QueryCompiler.cs:line 177
   at Microsoft.Data.Entity.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler) in D:\k\EntityFramework\src\EntityFramework.Core\Query\Internal\CompiledQueryCache.cs:line 32
   at Microsoft.Data.Entity.Query.Internal.QueryCompiler.CompileQuery[TResult](Expression query) in D:\k\EntityFramework\src\EntityFramework.Core\Query\Internal\QueryCompiler.cs:line 134
   at Microsoft.Data.Entity.Query.Internal.QueryCompiler.Execute[TResult](Expression query) in D:\k\EntityFramework\src\EntityFramework.Core\Query\Internal\QueryCompiler.cs:line 78
   at Microsoft.Data.Entity.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression) in D:\k\EntityFramework\src\EntityFramework.Core\Query\Internal\EntityQueryProvider.cs:line 37
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)

### PR DESCRIPTION
**Pull request by natemcmaster**
Fix #3633 

cc @maumar @anpete

### CODE REVIEW

**Comment by anpete** [comment]
@natemcmaster @maumar Not sure what happened to @maumar's original fix, but I prefer the approach in the origin PR. I.e. Enhance the QuerySourceTrackingExpressionVisitor to return null for this case.

**Comment by natemcmaster** [comment]
The PR @maumar had introduced a bug into other query tests. I'm going to rework this a little as there is a low-hanging, slight perf improvement available.

**Comment by natemcmaster** [comment]
Updated using @maumar 's original approach but with an additional condition check

File: src/EntityFramework.Core/Query/ExpressionVisitors/Internal/QuerySourceTracingExpressionVisitor.cs [hunk 1]
```diff
@@ -94,5 +95,16 @@ protected override Expression VisitSubQuery(SubQueryExpression expression)
         protected override Expression VisitLambda<T>(Expression<T> node) => node;
 
         protected override Expression VisitInvocation(InvocationExpression node) => node;
+        protected override Expression VisitMethodCall(MethodCallExpression node)
+        {
+            var referenceSource = node.Arguments.FirstOrDefault() as QuerySourceReferenceExpression;
+            if(node.Method.IsGenericMethod
```
Comment (anpete) [inline src/EntityFramework.Core/Query/ExpressionVisitors/Internal/QuerySourceTracingExpressionVisitor.cs]: Can we just prune all Property method calls?

Comment (natemcmaster) [inline src/EntityFramework.Core/Query/ExpressionVisitors/Internal/QuerySourceTracingExpressionVisitor.cs]: No. We need to leave some e.g. GroupJoin with shadow properties queries.

**Comment by anpete** [comment]
:shipit:

**Pull request merged**